### PR TITLE
Add Log4sly support.

### PR DESCRIPTION
**Pull request by hdasch**
Add support for Sly in addition to Slime.  This is offered as an alternative to pull request #21.  The discussion of pull request #21 suggested an alternative to simply renaming "slime" to "sly".  The change requires a bit more than introducing `backed-eval'.  But this version does support both Slime and Sly.

There are changes included to cope with the fact that Sly enables `font-lock' while Slime does not.  And the introduction of aliases may be controversial.  Feedback welcome.

This includes pull request #38 and #39, as I do not see a good way of just submitting log4sly support separately.  This pull requests does depend on pull request #39, but does not depend on pull request #38.

Guidance or correction gratefully accepted.

### CODE REVIEW

File: elisp/log4slime.el [hunk 1]
```diff
@@ -92,6 +95,202 @@ argument, the parent effective log level (string)")
...
+          (const :tag "slime" slime)
+          (const :tag "sly" sly)))
+
+(defmacro log4slime-with-backend-current-thread (value &rest body)
+  "Bind correct backend thread variable to VALUE, evaluate BODY."
+  (declare (debug (sexp &rest form))
+           (indent defun))
+  `(let ((,(pcase log4slime-backend
+            ('slime  (intern "slime-current-thread"))
+            ('sly  (intern "sly-current-thread")))
```
[the remark this answers is not in the document]
Comment (hdasch) [reply elisp/log4slime.el:118]: That is certainly broken.  The comma is a bug.  The conditional should be deferred to runtime.

`cl-case` makes as much, or more, sense here.  I'll convert the `pcase` to `cl-case`.

File: elisp/log4slime.el [hunk 2]
```diff
@@ -92,6 +95,202 @@ argument, the parent effective log level (string)")
...
+          ,value))
+    ,@body))
+
+(defun log4slime-backend-eval (expr &optional package
+                                    cancel-on-input cancel-on-input-retval)
+  "Evaluate EXPR on the superior backend lisp and return the result.
+For the Sly/Slynx backend, if CANCEL-ON-INPUT, cancel the request
```
[the remark this answers is not in the document]
Comment (hdasch) [reply elisp/log4slime.el:125]: It is, indeed.  Thanks for catching that.

File: elisp/log4slime.el [hunk 3]
```diff
@@ -1,6 +1,6 @@
 ;;; -*- Mode: Emacs-Lisp; -*-
 ;;;
-;;; Copyright (c) 2012, Max Mikhanosha. All rights reserved.
+;;; Copyright (c) 2012, 2021, Max Mikhanosha. All rights reserved.
```
[the remark this answers is not in the document]
Comment (hdasch) [reply elisp/log4slime.el:3]: Thanks.  Will do.

File: elisp/log4slime.el [hunk 4]
```diff
@@ -13,7 +13,10 @@
 ;;; See the License for the specific language governing permissions and
 ;;; limitations under the License.
 
-(require 'slime)
+(when (featurep 'slime)
+  (require 'slime))
+(when (featurep 'sly)
+  (require 'sly))
```
[the remark this answers is not in the document]
Comment (hdasch) [reply elisp/log4slime.el:19]: Indeed.  I misunderstood `featurep`.  I was trying to determine if the package was available, but not loaded.  I see that is not what `featurep` indicates.

Perhaps, a better approach would be `(require 'slime nil 'noerror)`, similarly for `sly`.

But your comment [below](https://github.com/sharplispers/log4cl/pull/40#discussion_r697383271) probably eliminates this altogether.  Followup below.

File: elisp/log4slime.el [hunk 5]
```diff
@@ -92,6 +95,202 @@ argument, the parent effective log level (string)")
...
+Return nil if there’s no connection.")
+
+(defun log4slime-backend-buffer-file-name ()
+  "Return the lisp file name associated with current buffer.
+If ‘slime-buffer-package’ or ‘sly-buffer-package’ (as
+appropriate) has a value then return that, otherwise search for
+and read an ‘in-package’ form."
+  (pcase log4slime-backend
+    ('slime (buffer-file-name))
+    ('sly (funcall (intern "sly-to-lisp-filename") (buffer-file-name)))))
```
[the remark this answers is not in the document]
Comment (hdasch) [reply elisp/log4slime.el:264]: Yes, the two backend APIs are different enough that it is hard to get clean/y account for differences.

It is all the harder if both backends may be loaded at the same time.

If I understand your suggestion correctly, I could introduce a base backend class to dispatch on.  Follow that with a derived backend specific class for each of `log4slime` and `log4sly`.  An instance of one or the other class needs to be registered with the common code.  Then convert the thirteen or so functions that differ to `defgeneric`, adding a reference to that instance variable.

Common code in `log4emacs.el` (or `log4common.el`, or whatever) would get the bulk of the code that is currently in `log4sime.el`.  And `log4slime.el` and `log4sly.el` would provide the `defmethods` and a `make-instance` of their respective classes to the common code.

Do I understand your suggestion correctly?

File: elisp/log4slime.el [hunk 6]
```diff
@@ -92,6 +95,202 @@ argument, the parent effective log level (string)")
...
+SINGLE-LOCATION is true if all xrefs point to the same location.
+FILE-ALIST is an alist of the form ((FILENAME . (XREF ...) ...).")
+
+(log4slime-define log4slime-backend-push-definition-stack ()
+  "Add point to the find-tag-marker-ring.")
+
+(log4slime-define log4slime-backend-xref.location (xref)
+  "Access slot \"location\" of backend XREF struct CL-X.")
+
+(log4slime-define log4slime-backend-length= (seq n)
```
[the remark this answers is not in the document]
Comment (hdasch) [reply elisp/log4slime.el:250]: I am not sure what you suggest I duplicate here.  The functions are defined differently in `slime` and `sly`.  Maybe this is not important given your next comment.  I will ask for clarification later, if I need it.

File: elisp/log4slime.el (hunk 5 above)
[the remark this answers is not in the document]
Comment (scymtym) [reply elisp/log4slime.el:264]: > Do I understand your suggestion correctly?

Yes. Do you think such an approach makes sense?

[the remark this answers is not in the document]
Comment (hdasch) [reply elisp/log4slime.el:264]: It does make sense.  And it seems a cleaner solution.

I am willing to do this.  But I have questions as to naming and review preparation before I get started.

As to naming:
- common code in log4cl.el, or would you prefer another name?
- should we keep the `log4slime-` prefix across all three files, or use three prefixes?  Emacs guidelines simply suggest consistency.  On the other hand, melpazoid complains if the prefix does not follow package and filename.

As to review preparation, I still am getting my head around the github workflow.  This will likely be a large changeset.  How can I break this up to make it easier to review?

[the remark this answers is not in the document]
Comment (scymtym) [reply elisp/log4slime.el:264]: > common code in log4cl.el, or would you prefer another name?

That seems fine as it matches the project name and the ASDF systems already use this naming scheme.

> should we keep the log4slime- prefix across all three files, or use three prefixes? Emacs guidelines simply suggest consistency. On the other hand, melpazoid complains if the prefix does not follow package and filename.

I don't have much experience with modern Emacs development. Keeping the prefix would be good for existing users, but I can see how that might violate the recommended style.

> This will likely be a large changeset. How can I break this up to make it easier to review?

Multiple self-contained commits help. For example, refactoring things to use a backend object and generic functions could be one commit, putting SLIME-specific functions into a separate file could be one commit, and adding SLY support could be one commit. Beyond that, I don't know how to make things easier to review.

[the remark this answers is not in the document]
Comment (hdasch) [reply elisp/log4slime.el:264]: > > should we keep the log4slime- prefix across all three files, or use three prefixes? Emacs guidelines simply suggest consistency. On the other hand, melpazoid complains if the prefix does not follow package and filename.
> 
> I don't have much experience with modern Emacs development. Keeping the prefix would be good for existing users, but I can see how that might violate the recommended style.

Hmm.  Consistency seems more important than trying to be to pedantic.  I will avoid unnecessary renames.  Fix it later if it seems an issue.  I suspect it will not.

> > This will likely be a large changeset. How can I break this up to make it easier to review?
> 
> Multiple self-contained commits help. For example, refactoring things to use a backend object and generic functions could be one commit, putting SLIME-specific functions into a separate file could be one commit, and adding SLY support could be one commit. Beyond that, I don't know how to make things easier to review.

Good.  I will structure this as I would a patch stream.  It will take a bit to refactor and test.

If you would like to close this pull request, I will open a new one what I have something to review.  If you leave it open, I will update it.

**Comment by hdasch** [comment]
I have pushed a version of my proposed refactoring that "works for me".  Nothing is as easy as it looks.  I needed to refactor the CL side (log4slime.lisp) as well as the elisp changes we discussed earlier.

I am not ready to request a detailed review.  Instead I have a few more questions.  First the questions, then a bit of expansion explaining why I ask.

- What is the earliest version of Emacs that this work needs to support?
- =log4sly.lisp= is the natural alternative to =log4slime.lisp= for the CL code.  Should this be =log4slynk.lisp= instead?
- Should version numbers be bumped to 1.1.4, 1.2.0, or 2.0.0?

For the question of Emacs versions, I used cl-defgeneric and cl-defmethod.  These are now preferred to defgeneric and defmethod, as the latter were declared obsolete since Emacs 25.1.  Is it acceptable to require Emacs version 25.1 or later?

For Linux users, Emacs 25.1 was delivered with Debian 10.  So this would break on Debian Stretch and earlier.  I have no idea what Emacs versions BSD users might be using.

Naming the CL interface Sly vs Slynk: my earlier patch blithely substituted "sly" for "slime" in =log4slime-check-connection= (which loads the CL code).  That worked for me because my =quicklisp= tree includes =ultralisp=.  The =(asdf:load-system :log4sly)= picked up a copy of =:log4sly= from the =40ants= team's tree.  So my earlier testing worked when it should have failed.  

I am new to CL development.  How far do I need to go to avoid package name collision.

Semantic versioning suggest this is not an insignificant change.  I think that rules out 1.1.4.  

I understand that the 1.0.x -> 1.1.x change occurred with the introduction of Log4Slime.  That could argue for 1.2.0.  But I have changed the elisp to CL API to deliver a backend parameter to dispatch against.  That could be regarded as a breaking change.  So 2.0.0? or 1.2.0?  Should this land in a branch rather than master?

I am still testing.  But I wanted to raise questions before settling on a final version of the PR.

Thanks.

**Comment by hdasch** [comment]
I may well be overstating the version issue.  The API breakage is internal, not external.  And the Emacs support is a small part of the overall project.  But it may be worth considering adding version verification to the Emacs code to ensure it is talking to the correct CL code.

**Comment by hdasch** [comment]
On second thought, ignore my comments about API breakage.  It occurred to me that with a little more glue, I do not need to change the elisp/CL interface.  So version 1.1.4?  And what are your thoughts regarding minimal Emacs version requirements and CL package name?

**Comment by scymtym** [comment]
> For the question of Emacs versions, I used cl-defgeneric and cl-defmethod. These are now preferred to defgeneric and defmethod, as the latter were declared obsolete since Emacs 25.1. Is it acceptable to require Emacs version 25.1 or later?

I assume that requiring Emacs 25.1 or newer should be fine.

> Naming the CL interface Sly vs Slynk: my earlier patch blithely substituted "sly" for "slime" in `log4slime-check-connection` (which loads the CL code). That worked for me because my `quicklisp` tree includes `ultralisp`. The `(asdf:load-system :log4sly)` picked up a copy of `:log4sly` from the `40ants` team's tree. So my earlier testing worked when it should have failed.

How about calling the ASDF systems `log4cl.log4slime` and `log4cl.log4sly`?

The packages are already called `log4cl.*` which should be fine.

> So version 1.1.4? 

I don't see a problem with that.

Not a review, but some hints:
* I would not repeat the generic function documentation string in method documentation strings.
* I would not turn `log4slime-length=` into backend generic function. It could be just be an ordinary function in the generic part of the ELisp code that happens to be identical to the function in the SLIME ELisp code (That is, maybe just copy and paste the SLIME function).

**Comment by hdasch** [comment]
Good.  Will post a new version after removing the API changes and incorporating your suggestions.

Thank you.

**Comment by hdasch** [comment]
I think the current version of this is ready for review.  Is there anything else here that you want me to change?

I have another pull request to submit once this is approved.  My branch `compile-warnings`, based on this pull request, has another `cl -> cl-lib` fix needed by Sly but not for Slime.  I can submit that pull request now of after your review, at your discretion.

Thanks,
Hugh

File: elisp/log4sly.el [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+;;; limitations under the License.
+
+(require 'cl-lib)
+(require 'log4cl)
+(require 'sly)
+
+(declare-function log4slime-highlight-log-message "log4cl.el")
+(declare-function slime-repl-emit "ext:slime-repl.el")
+
+(defvar log4slime-backend)              ; From log4cl.el
```
Comment (scymtym) [inline elisp/log4sly.el:24]: Why are the ones that come from `log4cl` needed? I thought `require` would enforce the compile/load order.

File: src/log4sly.lisp [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+;;;
+;;; Unless required by applicable law or agreed to in writing, software
+;;; distributed under the License is distributed on an "AS IS" BASIS,
+;;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;;; See the License for the specific language governing permissions and
+;;; limitations under the License.
+
+(defpackage #:log4cl.log4sly
+  (:nicknames #:log4sly)
+  (:use #:cl #:log4cl)
```
Comment (scymtym) [inline src/log4sly.lisp:19]: I would probably not `:use` `log4cl` here. I can't immediately see where it comes into play, so spelling out the qualified symbol name can't be required in too many places.

File: README.md [hunk 9]
```diff
@@ -1,66 +1,78 @@
+# Introduction
```
Comment (scymtym) [inline README.md:1]: Can you swap the final two commits, that is update the documentation, then bump the version. Seems as if the documentation is not part of the change otherwise.

File: README.org [hunk 10]
```diff
@@ -27,6 +27,8 @@ the 28.0 development tree.  See issue [[https://github.com/sharplispers/log4cl/i
 
 ** Enabling Log4Slime
 
+To enable Log4Sly run the following in your Lisp:
```
Comment (scymtym) [inline README.org:30]: Should this say "Log4Slime"?

File: README.org [hunk 11]
```diff
@@ -59,6 +61,35 @@ You can verify if log4slime is enabled, by looking for the Log4CL menu
...
 
+** Sly support
+
+As of version 1.1.4, Log4Slime supports Sly as well as Slime.  The
+minimal supported Emacs version is now 25.1.
+
+To enable Log4Sly run the following in your Lisp:
+
+#+BEGIN_SRC common-lisp
+(ql:quickload :log4ly)
```
Comment (scymtym) [inline README.org:72]: Typo in `log4ly`?

Comment (scymtym) [reply README.org:72]: Also, isn't the system name `log4cl.log4sly`?

File: README.org [hunk 12]
```diff
@@ -27,6 +27,8 @@ the 28.0 development tree.  See issue [[https://github.com/sharplispers/log4cl/i
 
 ** Enabling Log4Slime
 
+To enable Log4Sly run the following in your Lisp:
+
 #+BEGIN_SRC common-lisp
 (ql:quickload :log4slime)
```
Comment (scymtym) [inline README.org:33]: Shouldn't this be `log4cl.log4slime` given the previous changes?

File: elisp/log4slime.el [hunk 13]
```diff
@@ -14,13 +14,64 @@
...
 
-(require 'slime)
 (require 'cl-lib)
 (require 'log4cl)
+(require 'slime)
 
 (declare-function log4slime-highlight-log-message "log4cl.el")
 (declare-function slime-repl-emit "ext:slime-repl.el")
 
+(defvar log4slime-backend)              ; From log4cl.el
```
Comment (scymtym) [inline elisp/log4slime.el]: Why are the ones that come from `log4cl` needed? I thought `require` would enforce the compile/load order.

**Review by scymtym — CHANGES_REQUESTED** [review]
I think this is almost ready to merge. Thank you very much for all the work.

Final remarks in commit comments.

File: elisp/log4sly.el (hunk 7 above)
Comment (hdasch) [reply elisp/log4sly.el]: Hmm.  I cannot recall where I saw a warning about this.  Perhaps it was a third party linter.

Testing confirm these are not needed.

Removed.

File: src/log4sly.lisp (hunk 8 above)
Comment (hdasch) [reply src/log4sly.lisp]: Indeed, only two symbol references needed a package name.  Fixed.

Applied similar change to `log4slime.lisp`, where `:use` clause originated.

File: README.md (hunk 9 above)
Comment (hdasch) [reply README.md:1]: Certainly.

File: README.org (hunk 10 above)
Comment (hdasch) [reply README.org]: Yes, stupid copy and paste error.  Fixed.

File: README.org (hunk 12 above)
Comment (hdasch) [reply README.org]: Yes, was relying on package nickname.  I have removed the nickname from both `log4cl.log4slime` and `log4cl.log4sly`, with a minor modification to the elisp code to spell out the package name.

File: README.org (hunk 11 above)
Comment (hdasch) [reply README.org]: Yes.  Fixed.

**Comment by hdasch** [comment]
I have pushed a fixed version.  Thank you for your feedback.

**Review by scymtym — CHANGES_REQUESTED** [review]
Found one more thing.

File: elisp/log4sly.el [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+(require 'cl-lib)
+(require 'log4cl)
+(require 'sly)
+
+(cl-defmethod log4slime-eval ((backend (eql :log4sly)) form)
+  (when (log4slime-check-connection)
+    ;; I swear it something in slime-eval screws with point sometimes
+    (save-excursion
+      (cl-destructuring-bind (fn . args) form
+        (let ((slime-current-thread t)
```
[the remark this answers is not in the document]
Comment (hdasch) [reply elisp/log4sly.el]: Gah!  How did I miss that?  Many thanks for catching this.  I have pushed a fix.

One more thing.  Would you have a look at https://github.com/hdasch/log4cl/commit/6d5b454b9fd5c196bea6d4d491549a558cbe3bbd.  Do you want that included before the version number bump?

Thanks.

**Comment by scymtym** [comment]
Looks good to me. Thank you for your work on this and sorry for the delay.
